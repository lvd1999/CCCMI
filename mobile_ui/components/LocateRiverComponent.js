import React, {Component, useState} from 'react';
import {
  SafeAreaView,
  View,
  Text,
  StyleSheet,
  PermissionsAndroid,
  Platform,
  Button,
  Dimensions,
  TextInput,
} from 'react-native';

import GetLocation from 'react-native-get-location';
export default class LocateTest extends Component {
  constructor() {
    super();

    this.state = {
      location: {latitude: null, longitude: null},
      error: null,
      searchInput: '',
      openMap: false,
      region: '',
      locationStatus: '',
    };
    this.textInputContent = React.createRef();
  }

  /**
   *life cycle
   *
   * @memberof LocateTest
   */
  componentDidMount = async () => (
    () => {
      const requestLocationPermission = async () => {
        if (Platform.OS === 'ios') {
          this.getOneTimeLocation();
        } else {
          try {
            const granted = await PermissionsAndroid.request(
              PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
              {
                title: 'Location Access Required',
                message: 'This App needs to Access your location',
              },
            );
            if (granted === PermissionsAndroid.RESULTS.GRANTED) {
              //To Check, If Permission is granted
              this.getOneTimeLocation();
            } else {
              this.setState({locationStatus: 'Permission Denied'});
            }
          } catch (err) {
            console.warn(err);
          }
        }
      };
      requestLocationPermission();
      return () => {
        Geolocation.clearWatch(watchID);
      };
    },
    []
  );

  /**
   *life cycle
   *
   * @memberof LocateTest
   */
  componentDidUpdate = async () => (
    () => {
      const requestLocationPermission = async () => {
        if (Platform.OS === 'ios') {
          this.getOneTimeLocation();
          this.subscribeLocationLocation();
        } else {
          try {
            const granted = await PermissionsAndroid.request(
              PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
              {
                title: 'Location Access Required',
                message: 'This App needs to Access your location',
              },
            );
            if (granted === PermissionsAndroid.RESULTS.GRANTED) {
              //To Check, If Permission is granted
              this.getOneTimeLocation();
              this.subscribeLocationLocation();
            } else {
              this.setState({locationStatus: 'Permission Denied'});
            }
          } catch (err) {
            console.warn(err);
          }
        }
      };
      requestLocationPermission();
      return () => {
        Geolocation.clearWatch(watchID);
      };
    },
    []
  );

  /**
   *get time location once
   *
   * @memberof LocateTest
   */
  getOneTimeLocation = () => {
    console.log('getOneTimeLocation called');
    this.setState({locationStatus: 'Getting Location ...'});
    GetLocation.getCurrentPosition({
      enableHighAccuracy: true,
      timeout: 15000,
    })
      .then((position) => {
        this.setState({locationStatus: 'You are Here'});

        //getting the Longitude from the location json
        const currentLongitude = JSON.stringify(position.longitude);

        //getting the Latitude from the location json
        const currentLatitude = JSON.stringify(position.latitude);

        //Setting Longitude state
        this.setState({
          location: {latitude: currentLatitude, longitude: currentLongitude},
        });
      })
      .catch((error) => {
        this.setState({locationStatus: error.message});
      });
  };

  /**
   *
   * getInput from TextInput
   * @param {*} input text
   * @memberof LocateTest
   */
  getInput(text) {
    this.setState({...this.state, searchInput: text});
  }

  /**
   *Search River
   *
   * @memberof LocateTest
   */
  searchRiver() {
    alert(this.state.searchInput);
  }

  /**
   *render input value
   *
   * @return {*}
   * @memberof LocateTest
   */
  renderInput() {
    let width = Dimensions.get('window').width - 20;
    return (
      <>
        <TextInput
          placeholder="E.g.  Liffey or 74.168462, 45.9812645"
          placeholderTextColor="#FFFFFF"
          width={width}
          style={[styles.localInput]}
          onChangeText={(text) => this.getInput(text)}
          value={
            this.state.location.latitude &&
            this.state.location.longitude &&
            this.state.location.latitude + ',' + this.state.location.longitude
          }></TextInput>
      </>
    );
  }

  /**
   *render content
   *
   * @return {*}
   * @memberof LocateTest
   */
  render() {
    return (
      <>
        <SafeAreaView style={{flex: 1, backgroundColor: '#1B262C'}}>
          <View style={styles.container}>
            <Text style={[styles.fonts, styles.titleOne]}>First Step...</Text>
            <Text style={[styles.fonts, styles.titleTwo]}>
              Locate the river
            </Text>
          </View>

          <View style={{alignItems: 'center'}}>{this.renderInput()}</View>
          <Text style={styles.textRemind}>River name or Coordinates</Text>
          <View style={{marginTop: 50}}>
            <Button
              title="search"
              color="green"
              onPress={() => this.searchRiver()}></Button>
            <Button
              title="Locate"
              onPress={() => this.getOneTimeLocation()}></Button>
          </View>
          <View>
            <Text style={styles.fonts}>{this.state.locationStatus}</Text>
            <Text style={styles.fonts}>
              Longitude: {this.state.location.longitude}
            </Text>
            <Text style={styles.fonts}>
              Latitude: {this.state.location.latitude}
            </Text>
          </View>
        </SafeAreaView>
      </>
    );
  }
}

/**
 * styles
 * @type {*}
 * */
const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
    alignItems: 'center',
    justifyContent: 'center',
  },
  boldText: {
    fontSize: 25,
    marginVertical: 16,
  },
  fonts: {
    color: '#FFFFFF',
    fontSize: 30,
  },
  container: {
    marginTop: 100,
    paddingLeft: 50,
  },
  titleOne: {
    fontWeight: 'bold',
  },
  titleTwo: {
    paddingLeft: 50,
    marginTop: 50,
  },
  textRemind: {
    marginTop: 5,
    paddingLeft: 30,
    color: '#FFFFFF',
  },
  localInput: {
    marginTop: 50,
    borderWidth: 1,
    height: 40,
    borderColor: '#95989A',
    borderRadius: 5,
    fontSize: 15,
    color: '#FFFFFF',
    paddingLeft: 10,
  },
});
