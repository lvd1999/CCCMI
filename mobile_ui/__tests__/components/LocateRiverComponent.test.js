import 'react-native';
import React from 'react';
import LocateRiverComponent from '../../components/LocateRiverComponent';
import renderer from 'react-test-renderer';

let findElement = function (tree, element) {
  let result = undefined;
  for (node in tree.children) {
    if (tree.children[node].props.testId == element) {
      result = true;
    }
  }
  return result;
};
describe.only('Locate river component test', () => {
  test('should render locateRiverComponent', () => {
    const component = renderer.create(<LocateRiverComponent />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should check default state test', () => {
    const component = renderer.create(<LocateRiverComponent />).getInstance();
    expect(component.state.error).toEqual(null);
    const location = {latitude: null, longitude: null};
    expect(component.state.location).toEqual(location);
    expect(component.state.searchInput).toEqual('');
    expect(component.state.locationStatus).toEqual('');
  });

  test('should get textInput input value', () => {
    const component = renderer.create(<LocateRiverComponent />).getInstance();
    component.getInput('this is a test text');
    expect(component.state.searchInput).toEqual('this is a test text');
  });

  test('should search input function', () => {
    const component = renderer.create(<LocateRiverComponent />).getInstance();
    component.searchInput = 'this is a test text';
  });

  test('should search input function', () => {
    const component = renderer.create(<LocateRiverComponent />).getInstance();
    component.searchInput = 'this is a test text';
  });

  test('should find user text input element', () => {
    const component = renderer.create(<LocateRiverComponent />).toJSON();
    expect(findElement(component, 'userTestInput')).toBeDefined;
  });
});
