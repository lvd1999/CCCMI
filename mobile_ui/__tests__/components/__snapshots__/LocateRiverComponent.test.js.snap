// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Locate river component test renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 100,
        "paddingLeft": 50,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1B262C",
            "fontSize": 30,
          },
          Object {
            "fontWeight": "bold",
          },
        ]
      }
    >
      First Step...
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#1B262C",
            "fontSize": 30,
          },
          Object {
            "marginTop": 50,
            "paddingLeft": 50,
          },
        ]
      }
    >
      Locate the river
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="E.g.  Liffey or 53.3539, -6.3557"
      placeholderTextColor="#00000"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderColor": "#95989A",
            "borderRadius": 5,
            "borderWidth": 1,
            "color": "#1B262C",
            "fontSize": 15,
            "height": 40,
            "marginTop": 50,
            "paddingLeft": 10,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
      width={730}
    />
  </View>
  <Text
    style={
      Object {
        "color": "#1B262C",
        "marginTop": 5,
        "paddingLeft": 30,
      }
    }
  >
    River name or Coordinates
  </Text>
  <View
    style={
      Object {
        "marginTop": 50,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "green",
              },
            ]
          }
        >
          search
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Locate
        </Text>
      </View>
    </View>
  </View>
  <View>
    <Text
      style={
        Object {
          "color": "#1B262C",
          "fontSize": 30,
        }
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "color": "#1B262C",
          "fontSize": 30,
        }
      }
    >
      Latitude: 
    </Text>
    <Text
      style={
        Object {
          "color": "#1B262C",
          "fontSize": 30,
        }
      }
    >
      Longitude: 
    </Text>
  </View>
</RCTSafeAreaView>
`;

exports[`Locate river component test should render locateRiverComponent 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 100,
        "paddingLeft": 50,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1B262C",
            "fontSize": 30,
          },
          Object {
            "fontWeight": "bold",
          },
        ]
      }
    >
      First Step...
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "#1B262C",
            "fontSize": 30,
          },
          Object {
            "marginTop": 50,
            "paddingLeft": 50,
          },
        ]
      }
    >
      Locate the river
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      placeholder="E.g.  Liffey or 53.3539, -6.3557"
      placeholderTextColor="#00000"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "borderColor": "#95989A",
            "borderRadius": 5,
            "borderWidth": 1,
            "color": "#1B262C",
            "fontSize": 15,
            "height": 40,
            "marginTop": 50,
            "paddingLeft": 10,
          },
        ]
      }
      testID="userTestInput"
      underlineColorAndroid="transparent"
      value=""
      width={730}
    />
  </View>
  <Text
    style={
      Object {
        "color": "#1B262C",
        "marginTop": 5,
        "paddingLeft": 30,
      }
    }
  >
    River name or Coordinates
  </Text>
  <View
    style={
      Object {
        "marginTop": 50,
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
              Object {
                "color": "green",
              },
            ]
          }
        >
          search
        </Text>
      </View>
    </View>
    <View
      accessibilityRole="button"
      accessibilityState={Object {}}
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Locate
        </Text>
      </View>
    </View>
  </View>
  <View>
    <Text
      style={
        Object {
          "color": "#1B262C",
          "fontSize": 30,
        }
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "color": "#1B262C",
          "fontSize": 30,
        }
      }
    >
      Latitude: 
    </Text>
    <Text
      style={
        Object {
          "color": "#1B262C",
          "fontSize": 30,
        }
      }
    >
      Longitude: 
    </Text>
  </View>
</RCTSafeAreaView>
`;
